internal/controller/ingressreplica_controller.go

```diff
@@ -0,0 +1,142 @@
+		return ctrl.Result{}, err
+	}
+
+	sha, err := utils.HashObject(ingressReplica.Spec)
+	if err != nil {
+		logger.Error(err, "failed to hash IngressReplica.Spec")
+		utils.MarkCondition(ingressReplica.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionFalse, v1alpha1.ReadyConditionReason, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	newIngress := &networkv1.Ingress{}
+	if err := r.Get(ctx, k8sClient.ObjectKey{Name: ingressReplica.Name, Namespace: ingressReplica.Namespace}, newIngress); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.Error(err, "failed to get new Ingress")
+			utils.MarkCondition(ingressReplica.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionFalse, v1alpha1.ReadyConditionReason, err.Error())
+			return ctrl.Result{}, err
+		}
+
+		newIngress = genIngress(ingressReplica, oldIngress)
+		if err := r.Client.Create(ctx, newIngress); err != nil {
+			logger.Error(err, "failed to create new Ingress")
+			utils.MarkCondition(ingressReplica.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionFalse, v1alpha1.ReadyConditionReason, err.Error())
+			return ctrl.Result{}, err
+		}
+		ingressReplica.Status.SHA = &sha
+		utils.MarkCondition(ingressReplica.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionTrue, v1alpha1.ReadyConditionReason, "")
+		return ctrl.Result{}, nil
+	}
+
+	// update a new ingress
+	if !ingressReplica.Status.IsSHAEqual(sha) {
+		updateIngress(ingressReplica, newIngress, oldIngress)
+		if err := r.Client.Update(ctx, newIngress); err != nil {
+			logger.Error(err, "failed to update new Ingress")
+			utils.MarkCondition(ingressReplica.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionFalse, v1alpha1.ReadyConditionReason, err.Error())
+			return ctrl.Result{}, err
+		}
+	}
+	ingressReplica.Status.SHA = &sha
+	utils.MarkCondition(ingressReplica.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionTrue, v1alpha1.ReadyConditionReason, "")
+	return ctrl.Result{}, reterr
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *IngressReplicaReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.IngressReplica{}).
+		Complete(r)
+}
+
+func genIngress(ingressReplica *v1alpha1.IngressReplica, oldIngress *networkv1.Ingress) *networkv1.Ingress {
+	newIngress := &networkv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        ingressReplica.Name,
+			Namespace:   ingressReplica.Namespace,
+			Labels:      oldIngress.Labels,
+			Annotations: oldIngress.Annotations,
+		},
+		Spec: networkv1.IngressSpec{
+			IngressClassName: oldIngress.Spec.IngressClassName,
+			DefaultBackend:   oldIngress.Spec.DefaultBackend,
+		},
+	}
+	updateIngress(ingressReplica, newIngress, oldIngress)
+	return newIngress
+}
+
+func updateIngress(ingressReplica *v1alpha1.IngressReplica, newIngress *networkv1.Ingress, oldIngress *networkv1.Ingress) {
```

Comment:
I think we should have some mechanism to merge in new annotations.  I'm on the fence as to whether we should simply take any annotation from the ingress replica itself and override the annotations on the old ingress or have an attribute on the spec.  Sort of leaning towards the first option.

Reply:
I have used `lo.Assign` method which merges multiple maps from left to right.
```
newIngress.Annotations = lo.Assign(newIngress.Annotations, oldIngress.Annotations, ingressReplica.Annotations)
```